src/theory/decision_manager.h

```diff
@@ -0,0 +1,118 @@
+/*********************                                                        */
+/*! \file decision_manager.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief decision_manager
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef __CVC4__THEORY__DECISION_MANAGER__H
+#define __CVC4__THEORY__DECISION_MANAGER__H
+
+#include <map>
+#include "theory/decision_strategy.h"
+
+namespace CVC4 {
+namespace theory {
+
+/** DecisionManager
+ *
+ * This class manages all "decision strategies" for theory solvers in
+ * TheoryEngine. A decision strategy is a callback in the SAT solver for
+ * imposing its next decision. This is useful, for instance, in
+ * branch-and-bound algorithms where we require that the first decision
+ * is a bound on some quantity. For instance, finite model finding may impose
+ * a bound on the cardinality of an uninterpreted sort as the first decision.
+ *
+ * This class maintains a user-context-dependent set of pointers to
+ * DecisionStrategy objects, which implement indivdual decision strategies.
+ *
+ * Decision strategies may be registered to this class via registerStrategy
+ * at any time during solving. They are cleared via a call to reset during
+ * TheoryEngine's postSolve method.
+ *
+ * Decision strategies have a fixed order, which is managed by the enumeration
+ * type StrategyId, where strategies with smaller id have higher precedence
+ * in our global decision strategy.
+ */
+class DecisionManager
+{
+ public:
+  enum StrategyId
+  {
+    // The order of the global decision strategy used by the TheoryEngine
+    // for getNextDecision.
+
+    //----- assume-feasibile strategies
+    //  These are required to go first for the sake of model-soundness. In
+    //  other words, if these strategies did not go first, we might answer
+    //  "sat" for problems that are unsat.
+    STRAT_QUANT_CEGQI_FEASIBLE,
+    STRAT_QUANT_SYGUS_FEASIBLE,
+    STRAT_QUANT_SYGUS_STREAM_FEASIBLE,
```

Comment:
It does matter who goes first.  Some of the dependencies I can think of, UF cardinality needs to come before bounded integers or strings, and sygus enumerator size must come before sygus unif number of enumerators. I'm not sure about all of the dependencies (I'm still surprised about the combinations of features that users request at times), but I do know that this order is safe, since notice that this enumeration type reflects the current order.